### PR TITLE
add functions to inspect model and optimizer status to trainer.py

### LINKED ISSUE #29016: Trainer: Functions to inspect model and optimizer status
### Feature request

In huggingface Trainer, are there any functions to inspect model and optimizer status? such as, how many parameters require grad, learning rate of each parameter, which optimizer group each parameter belong...

I didn't find any related function in Trainer, and I know implementing it by myself is easy, but I just want to know whether such functions already exist.

### Motivation

Such inspection is useful for correcting training.

### Your contribution

I propose a question.

### PR DESCRIPTION
**Pull request by CKeibel**
# What does this PR do?

Add functions `get_num_trainable_parameters` to return the number of parameters which require grad, `get_learning_rates` to return the learning rates of parameter groups and `get_optimizer_group` to return optimizer groups for parameters.

Fixes #29016 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts @muellerzr

### CODE REVIEW

**Review by muellerzr — COMMENTED** [review]
Very nice PR! Any chance you could add some basic tests in `tests/trainer/test_trainer.py` for these?

**Comment by CKeibel** [comment]
> Very nice PR! Any chance you could add some basic tests in `tests/trainer/test_trainer.py` for these?

Of course, I wasn't sure where to add them, but now I see that I missed that `test_trainer.py` exists. I will add the tests later. :)

**Comment by muellerzr** [comment]
Great, thanks! I'll give a ✅ after, great job and very clean implementation :)

**Review by amyeroberts — COMMENTED** [review]
Nice - thanks for adding! 

+1 for adding tests. Once those are done, you can ping for a final review and we can merge 🤗

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -3769,3 +3769,37 @@ def test_hyperparameter_search_backends(self):
...
         )
+
+
+class OptimizerAndModelInspectionTest(unittest.TestCase):
+    def test_get_num_trainable_parameters(self):
+        in_features = 128
+        out_features = 64
+        # in_features * out_features + bias
+        expected_num_params = in_features * out_features + out_features
+        model = nn.Sequential(nn.Linear(in_features, out_features))
```
Comment (muellerzr) [inline tests/trainer/test_trainer.py:3780]: Let's perhaps make this two linear layers, one that's frozen, one that's not, so this way trainable checks if we have different optimizer groups? :)

Comment (CKeibel) [reply tests/trainer/test_trainer.py:3780]: Yes, you're right, that would definitely be better :)

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -3769,3 +3769,42 @@ def test_hyperparameter_search_backends(self):
...
+            with self.assertRaises(ValueError):
+                trainer.get_optimizer_group()
+            trainer.create_optimizer()
+            # Get groups
+            num_groups = len(trainer.get_optimizer_group())
+            self.assertEqual(num_groups, 2)
+            # Get group of parameter
+            param = next(model.parameters())
+            group = trainer.get_optimizer_group(param)
+            self.assertIn(param, group["params"])
```
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:3810]: We use `@require_xxx` decorators to control this. This makes it possible to add several requirements without having many indents

```suggestion
@require_torch
class OptimizerAndModelInspectionTest(unittest.TestCase):
    def test_get_num_trainable_parameters(self):
        model = nn.Sequential(nn.Linear(128, 64), nn.Linear(64, 32))
        # in_features * out_features + bias
        layer_1 = 128 * 64 + 64
        layer_2 = 64 * 32 + 32
        trainer = Trainer(model=model)
        self.assertEqual(trainer.get_num_trainable_parameters(), layer_1 + layer_2)
        # Freeze the last layer
        for param in model[-1].parameters():
            param.requires_grad = False
        self.assertEqual(trainer.get_num_trainable_parameters(), layer_1)

    def test_get_learning_rates(self):
        model = nn.Sequential(nn.Linear(128, 64))
        trainer = Trainer(model=model)
        with self.assertRaises(ValueError):
            trainer.get_learning_rates()
        trainer.create_optimizer()
        self.assertEqual(trainer.get_learning_rates(), [5e-05, 5e-05])

        def test_get_optimizer_group(self):
            model = nn.Sequential(nn.Linear(128, 64))
            trainer = Trainer(model=model)
            # ValueError is raised if optimizer is None
            with self.assertRaises(ValueError):
                trainer.get_optimizer_group()
            trainer.create_optimizer()
            # Get groups
            num_groups = len(trainer.get_optimizer_group())
            self.assertEqual(num_groups, 2)
            # Get group of parameter
            param = next(model.parameters())
            group = trainer.get_optimizer_group(param)
            self.assertIn(param, group["params"])
```

**Review by amyeroberts — APPROVED** [review]
Nice tests! Thanks for adding! 

Just a small nit on the torch condition - then we're good to merge!

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (CKeibel) [reply tests/trainer/test_trainer.py:3810]: Great, thank you very much. That makes perfect sense :)

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -3769,3 +3769,41 @@ def test_hyperparameter_search_backends(self):
...
```
Comment (muellerzr) [inline tests/trainer/test_trainer.py:3809]:
```suggestion
    def test_get_optimizer_group(self):
        model = nn.Sequential(nn.Linear(128, 64))
        trainer = Trainer(model=model)
        # ValueError is raised if optimizer is None
        with self.assertRaises(ValueError):
            trainer.get_optimizer_group()
        trainer.create_optimizer()
        # Get groups
        num_groups = len(trainer.get_optimizer_group())
        self.assertEqual(num_groups, 2)
        # Get group of parameter
        param = next(model.parameters())
        group = trainer.get_optimizer_group(param)
        self.assertIn(param, group["params"])
```

**Review by muellerzr — APPROVED** [review]
Thanks! Looks good to me as well, let's fix that spacing though 😉

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (CKeibel) [reply tests/trainer/test_trainer.py:3809]: Oh great, I missed that one :)

**Comment by CKeibel** [comment]
Is there anything else to do? :)  @amyeroberts @muellerzr

**Comment by CKeibel** [comment]
Excellent thanks, then I'll start looking for a new issue to work on. :)

**Comment by amyeroberts** [comment]
Looking forward to the future PRs :) Thanks again for contributing to improving the library!
